### ISSUE TITLE
Discrepancy Between Native Pandas Types And Dask Computed Pandas Types

### ISSUE THREAD

**Issue by andoni-garcia-fgp** [dask/dask#11225]
**Describe the issue**:

Currently I am seeing a really weird behavior in which object types in the Pandas DataFrame returned from Dask are being mutated during a `df[col].astype(str)` call despite `copy=True` being the default. This behavior is not present in natively instantiated Pandas frames, only after Dask converts results back into Pandas.

The underlying framework location that is causing the mutation is when [astype calls out to `lib.ensure_string_array`](https://github.com/pandas-dev/pandas/blob/2a9855b55b9912d336592569136d580ae0aa8209/pandas/core/dtypes/astype.py#L93-L95) which is not inspectable.

One hypothesis is that the raw Pandas frame is backed by an actual contiguous piece of memory, and when creating a copy during astype and then changing the type of the Series' values, the original frame is unaffected. However, the Dask frame is backed by a pointer to the data and so when creating a copy during astype and then changing the type of the Series' value both frames reflect the new data type. However, this is purely speculation from the observed behavior without detailed insight into how the backing Series is implemented.

**Minimal Complete Verifiable Example**:

```python
import datetime as dt
import dask
import dask.dataframe as dd
import pandas as pd

from dask.distributed import Client

TEST_DATA_PDF: pd.DataFrame = pd.DataFrame(
    [[dt.date(2024, 1, 1), "key1", "value1"]],
    columns=["date", "id", "value"],
)


def coerce_to_string(df: pd.DataFrame) -> pd.DataFrame:
    new_df = pd.DataFrame(index=df.index)
    for col in df.columns:
        new_df[col] = df[col].astype(str)
    return new_df


def test_raw_pdf() -> None:
    test_pdf = TEST_DATA_PDF.copy()
    assert type(test_pdf.date[0]) == dt.date

    coerced_pdf = coerce_to_string(test_pdf)
    assert type(test_pdf.date[0]) == dt.date
    assert type(coerced_pdf.date[0]) == str


def test_dask_pdf() -> None:
    dask.config.set({"dataframe.convert-string": False})
    dask.config.set(scheduler="synchronous")
    dask_client = Client(processes=False)

    test_pdf = TEST_DATA_PDF.copy()
    assert type(test_pdf.date[0]) == dt.date

    test_ddf = dd.from_pandas(TEST_DATA_PDF, npartitions=1)
    dask_test_pdf = dask_client.compute(test_ddf).result()
    assert type(dask_test_pdf.date[0]) == dt.date

    coerced_df = coerce_to_string(dask_test_pdf)
    assert type(test_pdf.date[0]) == dt.date
    assert type(coerced_df.date[0]) == str

    # This assert fails. Somehow the original df is being mutated by astype(str)
    assert type(dask_test_pdf.date[0]) == dt.date
```

**Anything else we need to know?**:

**Environment**:
All code is being run on an EC2 instance. I had a colleague confirm they could replicate this behavior.

- Dask version:  '2024.7.0'
- Pandas version: '2.1.4'
- Python version: '3.11.9'
- Operating System: 'Ubuntu 22.04.4 LTS'
- Install method (conda, pip, source): poetry

**Comment by phofl** [comment]
Thanks for your report.

Firstly, I would not recommend to use astype(str), that is just an alias for astype(object) at the moment. 

That said, the thing you are seeing seems like a pandas bug that was fixed on pandas 2.2 (I can reproduce on 2.1.4 but not on 2.2). I remember putting up a PR that fixed the ensure_string_array call.

**Comment by phofl** [comment]
I am closing this here since it's hard to patch for us and fixed on newer pandas
